### PR TITLE
fix(notifications): users are again unsubscribed when friendship and …

### PR DESCRIPTION
**Pull request by hypeJunction**
…membership are deleted

Plugins was not updated to use the new event name, leading to users
still receiving notifiations from groups they are not members of and
users they are no longer friends with

### CODE REVIEW

**Comment by hypeJunction** [comment]
Will add a simpletest

**Comment by hypeJunction** [comment]
It looks like our deprecation notice was ignored by many, I am adding a commit to log an error when someone tries to register an event handler by relationship name for core relationships.

File: engine/classes/Elgg/EventsService.php [hunk 1]
```diff
@@ -18,6 +18,16 @@ class EventsService extends \Elgg\HooksRegistrationService {
 	const OPTION_DEPRECATION_VERSION = 'deprecation_version';
 
 	/**
+	 * {@inheritdoc}
+	 */
+	public function registerHandler($name, $type, $callback, $priority = 500) {
+		if (in_array($name, ['create', 'update', 'delete']) && in_array($type, ['member', 'friend', 'member_of_site', 'attached'])) {
+			$this->logger->error("'$name, $type' event is no longer triggered. Update your event registration to use '$name, relationsip'");
```
Comment (mrclay) [inline engine/classes/Elgg/EventsService.php]: "relationsip"

File: engine/classes/Elgg/EventsService.php [hunk 2]
```diff
@@ -18,6 +18,16 @@ class EventsService extends \Elgg\HooksRegistrationService {
 	const OPTION_DEPRECATION_VERSION = 'deprecation_version';
 
 	/**
+	 * {@inheritdoc}
+	 */
+	public function registerHandler($name, $type, $callback, $priority = 500) {
+		if (in_array($name, ['create', 'update', 'delete']) && in_array($type, ['member', 'friend', 'member_of_site', 'attached'])) {
```
Comment (mrclay) [inline engine/classes/Elgg/EventsService.php]: Checking $type first might save a few cycles.

**Comment by mrclay** [comment]
LGTM.

File: mod/notifications/start.php [hunk 3]
```diff
@@ -106,19 +108,20 @@ function notifications_plugin_pagesetup() {
...
-	foreach($NOTIFICATION_HANDLERS as $method => $foo) {
-		remove_entity_relationship($user_guid, "notify{$method}", $object_guid);
+	
+	if (!in_array($relationship->relationship, ['member', 'friend'])) {
+		return;
+	}
+	
+	$methods = array_keys(_elgg_services()->notifications->getMethodsAsDeprecatedGlobal());
+	foreach($methods as $method) {
+		remove_entity_relationship($relationship->guid_one, "notify{$method}", $relationship->guid_two);
```
Comment (hypeJunction) [inline mod/notifications/start.php]: will use `elgg_remove_subscription()`

**Comment by hypeJunction** [comment]
Here is the script I have used to clean up the mess this has created:

```php
	$dbprefix = elgg_get_config('dbprefix');
	$methods = array_keys(_elgg_services()->notifications->getMethods());
	$groups = new \ElggBatch('elgg_get_entities', [
		'types' => 'group',
		'limit' => 0,
		'callback' => false,
	]);

	foreach ($groups as $group) {
		$bad_subscribers = new \ElggBatch('elgg_get_entities', [
			'types' => 'user',
			'limit' => 0,
			'wheres' => [
				// Not a member of a group
				"NOT EXISTS (SELECT 1
					FROM {$dbprefix}entity_relationships
						WHERE guid_two = $group->guid
							AND guid_one = e.guid
							AND relationship = 'member')",
				// But has an active subscription to group
				"EXISTS (SELECT 1
					FROM {$dbprefix}entity_relationships
						WHERE guid_two = $group->guid
							AND guid_one = e.guid
							AND relationship LIKE 'notify%')",


			],
			'callback' => false,
		]);
		$bad_subscribers->setIncrementOffset(false);

		foreach ($bad_subscribers as $subscriber) {
			foreach ($methods as $method) {
				elgg_remove_subscription($subscriber->guid, $method, $group->guid);
			}
		}
	}
```

**Pull request merged**